consensus/istanbul/backend/randao.go

```diff
@@ -0,0 +1,152 @@
+package backend
+
+import (
+	"bytes"
+	"math/big"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/klaytn/klaytn/accounts/abi/bind/backends"
+	"github.com/klaytn/klaytn/blockchain/system"
+	"github.com/klaytn/klaytn/blockchain/types"
+	"github.com/klaytn/klaytn/common"
+	"github.com/klaytn/klaytn/common/hexutil"
+	"github.com/klaytn/klaytn/consensus"
+	"github.com/klaytn/klaytn/crypto"
+	"github.com/klaytn/klaytn/crypto/bls"
+)
+
+// For testing without KIP-113 contract setup
+type BlsPubkeyProvider interface {
+	GetBlsPubkey(chain consensus.ChainReader, proposer common.Address, num *big.Int) (bls.PublicKey, error)
+	ResetBlsCache()
+}
+
+type ChainBlsPubkeyProvider struct {
```

Comment:
`ChainBlsPubkeyReader` seems more fit, but not strongly suggest changing this.

Reply:
Thanks for the suggestion but I prefer "Provider".